manta-accounting/src/key.rs

```diff
@@ -524,6 +523,71 @@ where
         })
     }
 
+    /// Returns a new [`ViewKeyTable`] for `self`.
+    #[inline]
+    pub fn view_key_table(self) -> ViewKeyTable<'h, H> {
+        ViewKeyTable::new(self)
+    }
+}
+
+/// View Key Table
+pub struct ViewKeyTable<'h, H>
+where
+    H: HierarchicalKeyDerivationScheme + ?Sized,
+{
+    /// Account Keys
+    keys: AccountKeysMut<'h, H>,
+
+    /// Pre-computed View Keys
+    view_keys: BTreeMap<KeyIndex, H::SecretKey>,
+}
+
+impl<'h, H> ViewKeyTable<'h, H>
+where
+    H: HierarchicalKeyDerivationScheme + ?Sized,
+{
+    /// View Key Buffer Maximum Size Limit
+    pub const VIEW_KEY_BUFFER_LIMIT: usize = 16 * (H::GAP_LIMIT as usize);
+
+    /// Builds a new [`ViewKeyTable`] over the account `keys`.
+    #[inline]
+    pub fn new(keys: AccountKeysMut<'h, H>) -> Self {
+        Self {
+            keys,
+            view_keys: Default::default(),
+        }
+    }
+
+    /// Returns the account keys associated to `self`.
+    #[inline]
+    pub fn into_keys(self) -> AccountKeysMut<'h, H> {
+        self.keys
+    }
+
+    /// Returns the view key for this account at `index`, if it does not exceed the maximum index.
+    ///
+    /// # Limits
+    ///
+    /// This function uses a view key buffer that stores the computed keys to reduce the number of
+    /// times a re-compute of the view keys is needed while searching. The buffer only grows past
+    /// the current key bounds with a call to [`find_index_with_gap`](Self::find_index_with_gap)
+    /// which extends the buffer by at most [`GAP_LIMIT`]-many keys per round. To prevent allocating
+    /// too much memory, the internal buffer is capped at [`VIEW_KEY_BUFFER_LIMIT`]-many elements.
+    ///
+    /// [`GAP_LIMIT`]: HierarchicalKeyDerivationScheme::GAP_LIMIT
+    /// [`VIEW_KEY_BUFFER_LIMIT`]: Self::VIEW_KEY_BUFFER_LIMIT
+    #[inline]
+    pub fn view_key(&mut self, index: KeyIndex) -> Option<&H::SecretKey> {
+        btree_map::get_or_mutate(&mut self.view_keys, &index, |map| {
+            let next_key = self.keys.view_key(index)?;
+            if map.len() == Self::VIEW_KEY_BUFFER_LIMIT {
+                btree_map::pop_last(map);
+            }
+            map.insert(index, next_key);
+            map.get(&index)
```

Comment:
`map.get()` returns `next_key` and takes some time to search the btree. Could we somehow avoid this search cost?

Reply:
Yeah, it might be possible to do `map.entry(index).or_insert(next_key)`. Let me try that.